### ISSUE TITLE
-Wundefined-internal should warn when an explicit template instantiation declaration appears outside system headers

### ISSUE THREAD

**Issue by ldionne** [llvm/llvm-project#52672]
The following code will result in a warning about `shrink_to_fit` having internal linkage but not being defined:

```shell
#!/usr/bin/env bash

cat <<EOF | clang++ -xc++ - -c -o string.o
template <class T>
struct basic_string {
    __attribute__((internal_linkage))
    void shrink_to_fit();
};

template <class T>
void basic_string<T>::shrink_to_fit() { }

template class basic_string<char>;
EOF

cat <<EOF | clang++ -xc++ - -c -o main.o
//# 1 "foo.cpp" 1 3
template <class T>
struct basic_string {
    __attribute__((internal_linkage))
    void shrink_to_fit();
};

template <class T>
void basic_string<T>::shrink_to_fit() { }
//# 10 "foo.cpp" 2

extern template class basic_string<char>;

int main() {
    basic_string<char> s;
    s.shrink_to_fit();
}
EOF

clang++ string.o main.o

rm string.o main.o
```

The output is:

```
<stdin>:5:10: warning: function 'basic_string<char>::shrink_to_fit' has internal linkage but is not defined [-Wundefined-internal]
    void shrink_to_fit();
         ^
<stdin>:16:7: note: used here
    s.shrink_to_fit();
      ^
1 warning generated.
```

However, uncommenting the preprocessor directives will make Clang think that `basic_string` is defined inside a system header, and the warning will be silenced. I believe Clang should warn despite `basic_string` being defined inside a system header because the explicit template instantiation declaration itself appears outside of a system header no matter what.

This issue actually hid a tricky bug in libc++ as I explain in [this StackOverflow answer](https://stackoverflow.com/a/70339311/627587).

**Comment by dwblaikie** [comment]
Not sure about the proposed change to clang's diagnostic - could work. Though probably requires some more work beyond using a different source location to classify the diagnostic (also I /think/ changing the source location for the diagnostic classification also changes the location of the diagnostic - though that might be suitable, pointing to the explicit instantiation declaration for the warning, rather than to the declaration in the header).

Though, for this particular instance, would it have been caught by building libc++ in such a way that the headers are not classified as system headers? It seems it'd be suitable for the developers of libc++ to see/use compiler warnings as normal, even though libc++ users would prefer not to see those warnings (in case they have even more warnings enabled for their own project for which libc++ isn't clean of issues - or newer compilers with new flags, etc).

**Comment by ldionne** [comment]
We already build/run the test suite with libc++ headers not being classified as system headers, but it doesn't help. This requires an explicit instantiation declaration of `basic_string` with a user-defined type.

**Comment by dwblaikie** [comment]
Maybe that's some missing test coverage for libc++, then? If it's a supported use case that's broken/being fixed in libc++, presumably a test will be added to/could've been written previously libc++ that'd fail (to compile (if the warning is enabled/the build is done with headers not classified as system headers) and/or link) if it's run without the fix (to mark `shrink_to_fit` as inline).

But yeah, there's a bunch of diagnostic experiences like this that aren't ideal - another someone posted recently was about `[[nodiscard]]` and `std::function<void()>` (pass a nodiscard function to that `std::function` type and a discarding call to a nodiscard function will be instantiated in a system header - and thus, no warning). Maybe there's a generalization here to check the instantiation stack in both cases, though this one (-Wundefined-internal) might be a bit quirkier than the more common pattern seen with `[[nodiscard]]`
